File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode], acc=0) -> int:
        if not root:
            return acc
        
        left_depth = self.maxDepth(root.left, acc + 1)
        right_depth = self.maxDepth(root.right, acc + 1)

        return max(left_depth, right_depth)

File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func helper(cur *TreeNode, acc int) int {
    if cur == nil {
        return acc
    }

    left := helper(cur.Left, acc + 1)
    right := helper(cur.Right, acc + 1)

    if left > right {
        return left
    }

    return right
}

func maxDepth(root *TreeNode) int {
    return helper(root, 0)
}